File: backend/src/models/QuyTrinhTuyenDung.js
const mongoose = require('mongoose');

const buocQuyTrinhSchema = new mongoose.Schema({
  thuTu: {
    type: Number,
    required: true
  },
  tenBuoc: {
    type: String,
    required: [true, 'Vui lòng nhập tên bước'],
    trim: true
  },
  moTa: {
    type: String,
    trim: true
  },
  batBuoc: {
    type: Boolean,
    default: true
  }
}, { _id: false });

const quyTrinhTuyenDungSchema = new mongoose.Schema({
  ten: {
    type: String,
    required: [true, 'Vui lòng nhập tên quy trình'],
    trim: true,
    maxlength: [200, 'Tên quy trình không được quá 200 ký tự']
  },
  moTa: {
    type: String,
    trim: true
  },
  cacBuoc: {
    type: [buocQuyTrinhSchema],
    required: true,
    validate: {
      validator: function(v) {
        return v && v.length > 0;
      },
      message: 'Quy trình phải có ít nhất 1 bước'
    }
  },
  macDinh: {
    type: Boolean,
    default: false
  },
  trangThai: {
    type: String,
    enum: ['Hoat_dong', 'Khong_hoat_dong'],
    default: 'Hoat_dong'
  },
  taoBoi: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'NguoiDung',
    required: true
  }
}, {
  timestamps: true,
  collection: 'quy_trinh_tuyen_dung'
});

// Index
quyTrinhTuyenDungSchema.index({ ten: 1 });
quyTrinhTuyenDungSchema.index({ macDinh: 1 });

// Virtual để đếm số bước
quyTrinhTuyenDungSchema.virtual('soBuoc').get(function() {
  return this.cacBuoc.length;
});

// Đảm bảo chỉ có 1 quy trình mặc định
quyTrinhTuyenDungSchema.pre('save', async function(next) {
  if (this.macDinh && this.isModified('macDinh')) {
    await this.constructor.updateMany(
      { _id: { $ne: this._id } },
      { macDinh: false }
    );
  }
  next();
});

const QuyTrinhTuyenDung = mongoose.model('QuyTrinhTuyenDung', quyTrinhTuyenDungSchema);

module.exports = QuyTrinhTuyenDung;

File: frontend/lib/models/quy_trinh_tuyen_dung.dart
class BuocQuyTrinh {
  final int thuTu;
  final String tenBuoc;
  final String? moTa;
  final bool batBuoc;

  BuocQuyTrinh({
    required this.thuTu,
    required this.tenBuoc,
    this.moTa,
    this.batBuoc = true,
  });

  factory BuocQuyTrinh.fromJson(Map<String, dynamic> json) {
    return BuocQuyTrinh(
      thuTu: json['thuTu'] ?? 0,
      tenBuoc: json['tenBuoc'] ?? '',
      moTa: json['moTa'],
      batBuoc: json['batBuoc'] ?? true,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'thuTu': thuTu,
      'tenBuoc': tenBuoc,
      'moTa': moTa,
      'batBuoc': batBuoc,
    };
  }
}

class QuyTrinhTuyenDung {
  final String id;
  final String ten;
  final String? moTa;
  final List<BuocQuyTrinh> cacBuoc;
  final bool macDinh;
  final String trangThai;
  final String taoBoi;
  final DateTime createdAt;
  final DateTime updatedAt;

  QuyTrinhTuyenDung({
    required this.id,
    required this.ten,
    this.moTa,
    required this.cacBuoc,
    this.macDinh = false,
    required this.trangThai,
    required this.taoBoi,
    required this.createdAt,
    required this.updatedAt,
  });

  factory QuyTrinhTuyenDung.fromJson(Map<String, dynamic> json) {
    return QuyTrinhTuyenDung(
      id: json['_id'] ?? '',
      ten: json['ten'] ?? '',
      moTa: json['moTa'],
      cacBuoc: (json['cacBuoc'] as List? ?? [])
          .map((e) => BuocQuyTrinh.fromJson(e))
          .toList(),
      macDinh: json['macDinh'] ?? false,
      trangThai: json['trangThai'] ?? 'Hoat_dong',
      taoBoi: json['taoBoi'] ?? '',
      createdAt: DateTime.parse(json['createdAt'] ?? DateTime.now().toIso8601String()),
      updatedAt: DateTime.parse(json['updatedAt'] ?? DateTime.now().toIso8601String()),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      '_id': id,
      'ten': ten,
      'moTa': moTa,
      'cacBuoc': cacBuoc.map((e) => e.toJson()).toList(),
      'macDinh': macDinh,
      'trangThai': trangThai,
      'taoBoi': taoBoi,
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
    };
  }

  int get soBuoc => cacBuoc.length;
  bool get isActive => trangThai == 'Hoat_dong';
}